### DISCUSSION TITLE
HTTP smuggling requests cause timeouts in resteasy-netty4

### DISCUSSION THREAD

**Discussion by txema-martinez-scopely** [resteasy/resteasy#4351] [Q&A]
Hi,

We’ve encountered a potential issue in the **resteasy-netty4** library when handling bot requests that use HTTP smuggling techniques. In our case, we were able to reproduce the issue with an HTTP header containing an ASCII control character.

Here’s what happens:

The **Netty** `HttpObjectDecoder` transitions into the `BAD_MESSAGE` state when it processes one of these smuggling requests.
The `RestEasyHttpRequestDecoder` then receives this result and responds with a `400 Bad Request`, which is expected behavior.
However, the response includes a `Connection: keep-alive` header, so any subsequent request on that same connection reaches the `HttpObjectDecoder` while it's still in a bad state. This causes the request to be ignored, leading to client timeouts.
This issue could pose a risk when using load balancers, as the connection in a bad state might be reused later for a legitimate client.

We’ve experimented with two potential fixes that might be applied in the `RestEasyHttpRequestDecoder`:

1. Sending a `Connection: close` response header, ensuring the client closes the connection after receiving the `400 Bad Request`.
2. Calling `HttpObjectDecoder.reset()` to clear the bad state, allowing the decoder to process subsequent requests properly.

Find attached a minimal example to reproduce the problem. Just start the server and run the client, which will send one smuggling request, and then the next legitimate request will cause a timeout.
[resteasy-netty-smuggler.zip](https://github.com/user-attachments/files/17281616/resteasy-netty-smuggler.zip)

We’d like to hear your thoughts on whether either of these solutions makes sense. If so, we’re happy to open a PR with the proposed fix.

Regards

**Comment by jamezp** [comment]
Thank you for reporting this. We will look into and get back to you. I'm not sure yet what the best fix is going to be.

**Reply by adrianeui** [reply]
Hi, do you have any update on this issue ?

**Reply by jamezp** [reply]
Not yet. I've not had a chance to look into this unfortunately. I've had a CVE filed for it, https://nvd.nist.gov/vuln/detail/CVE-2024-9622, but I've not had time to look for a fix.

@txema-martinez-scopely It looks like I completely forgot to reply. If you've got a fix, please feel free to submit it. I've filed https://issues.redhat.com/browse/RESTEASY-3553 for this.

I'll have a look this morning to see if there is a way to solve this as I need to do a release today for other reasons, but I can't guarantee a fix today.

**Reply by txema-martinez-scopely** [reply]
Yes, I can try to open a PR with alternative solution 1 (Sending a `Connection: close` response header), which seems to me the most conservative approach

**Reply by txema-martinez-scopely** [reply]
Opened #4407

**Reply by jamezp** [reply]
Excellent! Thank you @txema-martinez-scopely. I'm going to be a bit in and out the next couple of days so if I don't get to look at it this week, I will early next week.
